### PR TITLE
v2

### PR DESCRIPTION
**Pull request by nleitao**
-followed your suggestions (jianyun)
-improved the commenting

-since the waypoints_cb is only published once, new waypoints are now updated overtime the car position changes (cheers @namzat)

### CODE REVIEW

File: ros/src/twist_controller/dbw_node.py [hunk 1]
```diff
@@ -58,8 +58,19 @@ def __init__(self):
 
         # TODO: Subscribe to all the topics you need to
 
+
```
Comment (jiangjin) [inline ros/src/twist_controller/dbw_node.py:61]: could u remove the redundant empty line?

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 2]
```diff
@@ -26,28 +28,84 @@
...
+
+                self.final_waypoints_pub.publish(subset_waypoints)
+
 
     def waypoints_cb(self, waypoints):
         # TODO: Implement
+
+        #since /base_waypoints is only published once, the given_waypoints is only changed once as well.
+        #I think we only need the .waypoints, so I've discarded the .header
+        if self.given_waypoints is None:
```
Comment (jiangjin) [inline ros/src/waypoint_updater/waypoint_updater.py:104]: no need to check if given_waypoints is None?

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 3]
```diff
@@ -26,28 +28,84 @@
...
+
+        dl = lambda a, b: math.sqrt((a.x-b.x)**2 + (a.y-b.y)**2  + (a.z-b.z)**2)
+        distance=9999
+        index=None
+
+        self.pose=msg.pose.position
+
+        #to prevent situations where the simulator gives the /current_pose before providing the /base_points
+        if self.given_waypoints is None:
+            pass
```
Comment (jiangjin) [inline ros/src/waypoint_updater/waypoint_updater.py]: you can return here, so u don't need to have `else` clause

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 4]
```diff
@@ -26,28 +28,84 @@
...
+                    point_distance=dl(self.given_waypoints[i].pose.pose.position,self.pose)
+                    if point_distance<distance:
+                        distance=point_distance
+                        index=i
+                self.new_wp=index
+                self.old_wp=self.new_wp
+                self.old_pose=msg.pose
+
+                #to fix the case when the last waypoint is outside of the waypoints list
+                subset_end_index=self.new_wp+LOOKAHEAD_WPS
```
Comment (jiangjin) [inline ros/src/waypoint_updater/waypoint_updater.py:85]: where to define LOOKAHEAD_WPS?

Comment (namzat) [reply ros/src/waypoint_updater/waypoint_updater.py:85]: LOOKAHEAD_WPS is defined on line 26

Comment (jiangjin) [reply ros/src/waypoint_updater/waypoint_updater.py:85]: ah, that's my bad, please ignore this.

File: ros/src/waypoint_updater/waypoint_updater.py (hunk 2 above)
Comment (nleitao) [reply ros/src/waypoint_updater/waypoint_updater.py:104]: I think is better to keep it because if I removed there are some times that ROS returns an error

**Pull request closed without merging**